src/system/rpi_pico/system.c

```diff
@@ -0,0 +1,312 @@
+        *((uint8_t *)buf) = z_random_u8();
+    }
+}
+
+/*------------------ Memory ------------------*/
+void *z_malloc(size_t size) { return pvPortMalloc(size); }
+
+void *z_realloc(void *ptr, size_t size) {
+    _ZP_UNUSED(ptr);
+    _ZP_UNUSED(size);
+    // realloc not implemented in FreeRTOS
+    assert(false);
+    return NULL;
+}
+
+void z_free(void *ptr) { vPortFree(ptr); }
+
+#if Z_FEATURE_MULTI_THREAD == 1
+// In FreeRTOS, tasks created using xTaskCreate must end with vTaskDelete.
+// A task function should __not__ simply return.
+typedef struct {
+    void *(*fun)(void *);
+    void *arg;
+    EventGroupHandle_t join_event;
+} z_task_arg;
+
+static void z_task_wrapper(void *arg) {
+    z_task_arg *targ = (z_task_arg *)arg;
+    targ->fun(targ->arg);
+    xEventGroupSetBits(targ->join_event, 1);
+    vTaskDelete(NULL);
+}
+
+static z_task_attr_t z_default_task_attr = {
+    .name = "",
+    .priority = configMAX_PRIORITIES / 2,
+    .stack_depth = 5120,
+};
+
+/*------------------ Thread ------------------*/
+z_result_t _z_task_init(_z_task_t *task, z_task_attr_t *attr, void *(*fun)(void *), void *arg) {
+    z_task_arg *z_arg = (z_task_arg *)z_malloc(sizeof(z_task_arg));
+    if (z_arg == NULL) {
+        return -1;
+    }
+
+    z_arg->fun = fun;
+    z_arg->arg = arg;
+    z_arg->join_event = task->join_event = xEventGroupCreate();
+
+    if (attr == NULL) {
+        attr = &z_default_task_attr;
+    }
+
+    if (xTaskCreate(z_task_wrapper, attr->name, attr->stack_depth, z_arg, attr->priority, &task->handle) != pdPASS) {
+        return -1;
+    }
+
+    return 0;
+}
+
+z_result_t _z_task_join(_z_task_t *task) {
+    xEventGroupWaitBits(task->join_event, 1, pdFALSE, pdFALSE, portMAX_DELAY);
+    return 0;
+}
+
+z_result_t _z_task_detach(_z_task_t *task) {
+    _ZP_UNUSED(task);
+    assert(false);
+    return _Z_ERR_GENERIC;
+}
+
+z_result_t _z_task_cancel(_z_task_t *task) {
+    vTaskDelete(task->handle);
+    return 0;
+}
+
+void _z_task_free(_z_task_t **task) {
+    z_free((*task)->join_event);
+    z_free(*task);
+}
+
+/*------------------ Mutex ------------------*/
+z_result_t _z_mutex_init(_z_mutex_t *m) {
+    *m = xSemaphoreCreateRecursiveMutex();
+    return *m == NULL ? -1 : 0;
+}
+
+z_result_t _z_mutex_drop(_z_mutex_t *m) {
+    z_free(*m);
+    return 0;
+}
+
+z_result_t _z_mutex_lock(_z_mutex_t *m) { return xSemaphoreTakeRecursive(*m, portMAX_DELAY) == pdTRUE ? 0 : -1; }
+
+z_result_t _z_mutex_try_lock(_z_mutex_t *m) { return xSemaphoreTakeRecursive(*m, 0) == pdTRUE ? 0 : -1; }
+
+z_result_t _z_mutex_unlock(_z_mutex_t *m) { return xSemaphoreGiveRecursive(*m) == pdTRUE ? 0 : -1; }
+
+/*------------------ CondVar ------------------*/
+static UBaseType_t CONDVAR_MAX_WAITERS_COUNT = 255;
+
+z_result_t _z_condvar_init(_z_condvar_t *cv) {
+    if (!cv) {
+        return _Z_ERR_GENERIC;
+    }
+    cv->mutex = xSemaphoreCreateMutex();
+    cv->sem = xSemaphoreCreateCounting(CONDVAR_MAX_WAITERS_COUNT, 0);
```

Comment:
Why limiting the max count to this particular value ? Larger max count value does not affect the needed memory for the semaphore so we can safely use "unlimited" count:
```suggestion
    cv->sem = xSemaphoreCreateCounting((UBaseType_t)(~0), 0);
```